### PR TITLE
[9.0] report_py3o

### PR DESCRIPTION
**Pull request by faide**
This is the result of OCA days work in the "Open Document Reporting Workgroup"

We will add unit tests on this PR later on

The companion document is available here: https://docs.google.com/spreadsheets/d/1YgniNHjEWDHXZkhYWwasxmIklW734TL143euQ2Sjepc/

### CODE REVIEW

File: report_py3o/LICENSE
Comment (elicoidal) [inline report_py3o/LICENSE:1]: @faide   Thanks for your contribution to the OCA. 
Let me introduce to you some basic information about contribution in OCA and I will review other points in more comments inside the code..
In general, please follow the principles: 
- Contribution guidelines: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md
- Please follow the OCA README template:
  https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
- We recommend to include a screenshot as well to help new comers (nice to have)
- Please use simple headers in all py files following OCA templates
  https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

In particular here, this file should be removed (normally it is at repo level)

File: report_py3o/README.md
Comment (elicoidal) [inline report_py3o/README.md]: this should be transformed into README.rst template

File: report_py3o/__openerp__.py
Comment (elicoidal) [inline report_py3o/__openerp__.py]: remove =>README

File: report_py3o/security/ir.model.access.csv
Comment (elicoidal) [inline report_py3o/security/ir.model.access.csv]: should not you add a blank line?

**Review by elicoidal — CHANGES_REQUESTED** [review]
Thanks for the contribution.
some formal modification to fit the current OCA branding

**Comment by pedrobaeza** [comment]
Can you please squash some commits together that are not meaningful in the final repository?

**Comment by NL66278** [comment]
Just some general comments. Researching what to do to get py3o uno to run I see a lot of confusing and incomplete information, examples not working, docker images that are no longer there etc. It makes me doubt whether this will really be an improvement compared to aeroo reports.

**Comment by NL66278** [comment]
```
Summary of the commands to install the server on debian jessie (run as root):
# aptitude install docker.io
# pip install docker-compose
# wget https://bitbucket.org/xcgd/py3o.fusion/raw/76dcb04477c4ddcaa303db94ae340561c2da75a9/docker-compose.yml
# export COMPOSE_API_VERSION=1.18
# docker-compose up
```

**Comment by alexis-via** [comment]
@NL66278 I don't agree with you. The documentation, information, code quality, test suite of py3o is much better than the one of aeroo. Aeroo report doesn't really have any documentation at all, no unit tests, etc... Of course, as py3o is new for the vast majority of Odoo community members, those who plan to switch from aeroo to py3o will have to go through the learning curve, but it should be easier than aeroo for newcomers. If you came at the OCA code sprint and worked with the libreoffice-reporting group, you would have learnt many useful things, ask your questions to the py3o experts and participate in the choice of the OCA libreoffice-based reporting engine. But criticizing today on this PR when the work is still in progress is certainly not a good way to welcome a new contribution and thank the hard work of the authors of py3o.

**Comment by alexis-via** [comment]
I just succedded in generating my first py3o PDF document with odoo9 + report_py3o of this PR + py3o.fusion server + py3o.render + libreoffice on my Ubuntu 16.04... I'm so happy ! :) Big thanks to @faide for his help to make it work. I'll document the steps and tricks to configure all that (but after my presentation tomorrow on e-invoice and e-orders). :+1: once my PR will be merged in faide's branch https://github.com/faide/reporting-engine/pull/9

File: report_py3o/README.md
Comment (lasley) [inline report_py3o/README.md]: IMO LibreOffice/OpenOffice is too application specific. This module seems to convert into the [Open Document Format (ODF)](https://en.wikipedia.org/wiki/OpenDocument)

File: report_py3o/NEWS
Comment (lasley) [inline report_py3o/NEWS]: Please remove this file

File: report_py3o/__openerp__.py
Comment (lasley) [inline report_py3o/__openerp__.py]: Same note regarding the application specific name. This is converting into the Open Document Format, as there is no such thing as LibreOffice format. These same templates should also work just fine in Microsoft Office.

File: report_py3o/__openerp__.py
Comment (lasley) [inline report_py3o/__openerp__.py]: Add `Odoo Community Association (OCA)`

File: report_py3o/__openerp__.py
Comment (lasley) [inline report_py3o/__openerp__.py]: base is implicit, please remove

File: report_py3o/models/ir_report.py
Comment (lasley) [inline report_py3o/models/ir_report.py]: The name of this file should be the snake_case_version of the model name

File: report_py3o/models/ir_report.py
Comment (lasley) [inline report_py3o/models/ir_report.py]: The name of this class should be the UpperCamelCased version of the model name

File: report_py3o/models/ir_report.py
Comment (lasley) [inline report_py3o/models/ir_report.py]: api.one is deprecated

File: report_py3o/models/ir_report.py
Comment (lasley) [inline report_py3o/models/ir_report.py]: api.one is deprecated

File: report_py3o/models/ir_report.py
Comment (lasley) [inline report_py3o/models/ir_report.py]: Please add translation  to errors

File: report_py3o/models/ir_report.py
Comment (lasley) [inline report_py3o/models/ir_report.py]: Add translation

File: report_py3o/models/ir_report.py
Comment (lasley) [inline report_py3o/models/ir_report.py]: All external dependency imports should be guarded.

```
try:
    from py3o.formats import Formats
except ImportError:
    _logger.info('Py3o is not installed')
```

File: report_py3o/models/ir_report.py
Comment (lasley) [inline report_py3o/models/ir_report.py]: Field definitions should be at top of class

File: report_py3o/models/ir_report.py
Comment (lasley) [inline report_py3o/models/ir_report.py]: This is possible with ORM. Was there a specific reason that SQL was used?

File: report_py3o/models/py3o_server.py
Comment (lasley) [inline report_py3o/models/py3o_server.py:15]: Odoo convention for this field name is `active`, which will provide some special handling such as auto-hiding when deactivated

File: report_py3o/models/py3o_template.py
Comment (lasley) [inline report_py3o/models/py3o_template.py]: Same note re application specific text. This will work just fine in MS Office. Prefer the document format instead of the application name

File: report_py3o/models/py3o_server.py
Comment (lasley) [inline report_py3o/models/py3o_server.py:7]: IMO we're switching naming schemas all over the place, and not many people will know what `py3o` is. Please switch this to ODFServer or similar

File: report_py3o/models/py3o_template.py
Comment (lasley) [inline report_py3o/models/py3o_template.py:7]: IMO we're switching naming schemas all over the place, and not many people will know what py3o is. Please switch this to ODFTemplate or similar

File: report_py3o/py3o_parser.py
Comment (lasley) [inline report_py3o/py3o_parser.py]: Guard all external imports

File: report_py3o/py3o_parser.py
Comment (lasley) [inline report_py3o/py3o_parser.py:75]: I would spell out `At the moment` for clarity

File: report_py3o/py3o_parser.py
Comment (lasley) [inline report_py3o/py3o_parser.py:79]: This is mixing Doxygen and sphinx param format. Prefer sphinx:

```
:param report_obj:
```

Also definition is spelled wrong

File: report_py3o/py3o_parser.py
Comment (lasley) [inline report_py3o/py3o_parser.py:132]: Please use new api

File: report_py3o/py3o_parser.py
Comment (lasley) [inline report_py3o/py3o_parser.py:183]: This comment does not add anything to the meaning of the code, the raising of an error implicitly says you are telling the user about an error. No need to repeat yourself IMO

File: report_py3o/py3o_parser.py
Comment (lasley) [inline report_py3o/py3o_parser.py:203]: USe new api

File: report_py3o/views/ir_report.xml
Comment (lasley) [inline report_py3o/views/ir_report.xml]: You can remove this data tag

File: report_py3o/views/menu.xml
Comment (lasley) [inline report_py3o/views/menu.xml]: Remove data tag

File: report_py3o/views/py3o_server.xml
Comment (lasley) [inline report_py3o/views/py3o_server.xml]: Remove data tag

File: report_py3o/views/py3o_template.xml
Comment (lasley) [inline report_py3o/views/py3o_template.xml]: Remove data tag

**Review by lasley — CHANGES_REQUESTED** [review]
Thanks for the submission @faide - comments inline

Additionally and IMO, this module should be named `report_odf` - nobody is going to know what py3o is

**Comment by rvalyi** [comment]
guys, I'm not sure where it is the best to comment on this but in the comparison doc you said Aeroo has no Docker image. Well At Akretion we are using https://github.com/akretion/aeroocker in several projects for nearly a year now. That being said we are open to migrate to py3o in the future if it proves better.

File: report_py3o/models/ir_report.py
Comment (alexis-via) [reply report_py3o/models/ir_report.py]: No, api.one is not deprecated any more.

File: report_py3o/models/py3o_server.py
Comment (alexis-via) [reply report_py3o/models/py3o_server.py:7]: No, we won't change model names.

File: report_py3o/models/py3o_template.py
Comment (alexis-via) [reply report_py3o/models/py3o_template.py:7]: No, we keep py3o. The py3o project has existed for many years, it's name won't be changed.

File: report_py3o/models/ir_report.py
Comment (lmignon) [reply report_py3o/models/ir_report.py]: @lasley IMO in this case the model name is too long... ir_action_report_xml.py

File: report_py3o/models/py3o_template.py
Comment (lmignon) [reply report_py3o/models/py3o_template.py]: @lasley py3o_template refers to the file used as template... IMO the name appropriate

File: report_py3o/py3o_parser.py
Comment (lmignon) [reply report_py3o/py3o_parser.py:132]: @lasley here we are in the parser not in an odoo model. Therefore the new api is applicable and moreover our class extends a pure python class and we override only some methods... so we must keep the same method signature

**Comment by alexis-via** [comment]
I took into account most remarks of @lasley in my PR https://github.com/faide/reporting-engine/pull/9 against faide's branch.

**Comment by NL66278** [comment]
@alexis-via I am sorry that the doubts I raised are seen as just criticizing the hard work of people I certainly respect. It is nor was my intention just to criticise from the sidelines. But then the point  about having reviews is also to raise issues when you encounter them. As you should know I did come to the codesprint and talk to the py3o people, that I found very helpful. I made two bug reports on the py3o project on bitbucket complete with documentation on what I did for the sake of validation. And I wrote a later comment on how I got finally the py3o fusion server to work on Debian.  Later on I created a buildout recipe to create an Odoo instance with py3o reporting enabled. It is here: https://github.com/nl66278/reporting-engine/tree/buildout-9.0-testing. That said I think the documentation still leaves much to be desired. I am fully willing to participate in rectifying that. Just the raw work of what I found out yesterday I will attach to this comment. Actually I wrote up a lot more, but the parts were I was led astray by the existing documentation I deleted.

**Comment by NL66278** [comment]
WTF: github does not support ODF, while supporting proprietary formats like docx.
[2016-10-04-odoo-p3o-reporting.pdf](https://github.com/OCA/reporting-engine/files/511274/2016-10-04-odoo-p3o-reporting.pdf)

Well: is has difficulty with pdf, but here is my writeup of what to do to get the py3o fusion server working om debian.

**Comment by faide** [comment]
Hi Ronald I know documentation is still terse on the deployment parts but there are visible efforts and openness. Did you see the library usage documentation ?
http://py3otemplate.readthedocs.io
This is what we intend to have for the whole project.

I'll look into your issues and I am grateful we were able to discuss on them, I'll make sure to pin them down!

**Comment by NL66278** [comment]
Hello @faide yes I looked at that documentation that is certainly very useful. The part I was having an issue with was with the setup / installation part. If in any way my doubts seemed to apply a lack of appreciation for the great work done,  I would certainly apologize for that.

**Comment by alexis-via** [comment]
@NL66278 Thanks for your answer. Glad to see that we are now all working on improving py3o to have the best code and a good doc.

**Comment by lmignon** [comment]
@faide Can you provide a CLA for the following commiters?  http://odoo-community.org/page/website.cla

```
Vincent Lhote-Hatakeyama (no github login found)
@anael-lorimier (login unknown in OCA database)
@brmasson (login unknown in OCA database)
@faide (login unknown in OCA database)
@alexbrun (login unknown in OCA database)
@szekawong (login unknown in OCA database)
```

**Comment by lasley** [comment]
Rebuilding Runbot now that dependencies have been installed

**Comment by chermed** [comment]
Can someone explain to us : what is the difference between `Aeroo` And `py3o` ?

**Comment by lmignon** [comment]
@chermed Basically the 2 modules provides the same functionalities... (see https://docs.google.com/spreadsheets/d/1YgniNHjEWDHXZkhYWwasxmIklW734TL143euQ2Sjepc/edit#gid= for details)
Nevertheless thanks to XCG report_py3o is proposed to be included under the OCA umbrella. Besides being a clean and well-tested module, the inclusion under the OCA umbrella ensures that the module will continue to evolve across the Odoo versions and the community will have the opportunity to contribute to its development.

File: report_py3o/models/ir_report.py
Comment (bealdav) [reply report_py3o/models/ir_report.py]: guidelines seems says .debug instead of .info

**Comment by alexis-via** [comment]
I made a PR with the installation procedure in README https://github.com/faide/reporting-engine/pull/12

**Comment by alexis-via** [comment]
For me, this PR should be merged now (if the CLAs have been signed).

**Comment by alexis-via** [comment]
@faide During the code sprint, we managed to call the display_address() method in py3o with

```
py3o://function="display_address(o.partner_id)"
```

but the carriage returns were not taken into account in the ODT file. For example, the current result is:

```
35B rue Montgolfier 69100 VILLEURBANNE France
```

with everything on the same line. Do you have an idea how we could fix that ? It makes the method display_address almost unusable in py3o for the moment...

**Comment by eLBati** [comment]
Thanks for this work.

I am testing it and wondering how to avoid an empty field to be printed as **False**
How can I print an empty string?

**Review by lasley — APPROVED** [review]

**Comment by lasley** [comment]
Changes LGTM. 

Once @elicoidal confirms good as well, can you please squash the logical commits - specifically making sure to change the commit message on the one that tags me (2fbbdc9). Commit messages that tag users will spam them indefinitely 😦

**Comment by eLBati** [comment]
Thanks  @alexis-via  for the answer  https://twitter.com/alexisdlattre/status/791750379377754113

For now, we can anyway use the `if` py3o syntax
:+1:

**Comment by JosDeGraeve** [comment]
Hi,

I have tested the p3o reporting engine and git it up and running in practically no time.  Good work !

For reports there is a flag "is_local_fusion".  I'm a bit confused by this.  In order to generate a PDF report, an external py3o server is needed for 'fusion/renderserver/oooserver'.  However, with an external fusion, it is not possible to call odoo methods: that is a big restriction.

Why is it not possible to have the fusion part done on the odoo side + have the conversion step to PDF or other formats done by the rendering server ?

A feature that aeroo reports had, was to generate multiple reports separately.  This allowed, for example, to print multiple invoices in a single run, but with each invoice generated in a separate document.  If you need per-document page numbering, or want to eliminate nasty page alignment issues between documents, this feature comes very handy.  Would it be hard to implement such a feature ? 

Last question is how are translations handled ?  I can't find any info on that.

Tnx

**Comment by lmignon** [comment]
@josDeGraeve the flag 'is_local_fusion' is to use if you want to be able to call methods from your templates. The rendition and the fusion are 2 separates steps. Even if you select the local fusion you can specify a py3o server to render your template into pdf or doc.

**Comment by scigghia** [comment]
hi, i've read documentation and source, but i need an example report module to study 
is there?

**Comment by JosDeGraeve** [comment]
@scigghia for documentation, take a look at http://py3otemplate.readthedocs.io/en/latest/
As small sample doc can be found in  `demo/res_user.odt`

**Comment by lmignon** [comment]
@scigghia @JosDeGraeve You can also take a look at https://bitbucket.org/faide/py3o.template/src/1e5efa93e8e64a642174c129c63f3735c9be3e6a/example/?at=default

**Comment by lmignon** [comment]
A refactoring is proposed https://github.com/OCA/reporting-engine/pull/78#issuecomment-267570835

**Pull request closed without merging**
